packages/databus/src/index.ts

```diff
@@ -0,0 +1,135 @@
+/*-----------------------------------------------------------------------------
+| Copyright (c) Jupyter Development Team.
+| Distributed under the terms of the Modified BSD License.
+|----------------------------------------------------------------------------*/
+
+import { DisposableDelegate, IDisposable } from '@phosphor/disposable';
+
+import { ArrayExt } from '@phosphor/algorithm';
+
+import { ISignal, Signal } from '@phosphor/signaling';
+
+/**
+ * An interface for an abstract dataset, with a mimetype, data, and metadata.
+ *
+ * This is templated over the type of data `T` and the type of metadata `U`.
+ *
+ * #### Notes
+ * This interface is similar conceptually to that of nbformat.IMimeBundle,
+ * but with only a single mimetype and different structure. We expect there
+ * to be utilities to convert between this formats.
+ */
+export interface IDataset<T, U> {
+  /**
+   * The string mimetype for the dataset.
+   */
+  mimeType: string;
+
+  /***
+   * The data of the dataset.
+   *
+   * #### Notes
+   * The notion of data in the databus is completely abstract, and typed as an
+   * any. Publishers and consumers of a given mimetype will need to agree
+   * what the actual data is and cast it appropriately before use.
+   */
+  data: T;
+
+  /**
+   * The metadata associate with the dataset.
+   *
+   * #### Notes
+   * Like the data itself, the databus notion of metadata is entirely abstract.
+   * We are currently exploring different metadata schemes, so this may change
+   * in the future.
+   */
+  metadata: U;
+}
+
+/**
+ * A databus object, for managing datasets in JupyterLab.
+ */
+export class DataBus {
```

Comment:
What did you think of the discussion with @BoPeng about the naming of this? That is, we aren't really talking about shuttling data around, so much as registering and publishing it.

Reply:
Perhaps we can call items to be managed as `DataToken`, and managers in line of `DataRegistry` and `DataPublisher`?

Reply:
I am completely open to other names. I think `DataRegistry` is probably pretty neutral and descriptive.